planner/core/exhaust_physical_plans.go

```diff
@@ -1281,27 +1303,23 @@ func (p *LogicalJoin) tryToGetIndexJoin(prop *property.PhysicalProperty) (indexJ
 	forceRightOuter := inljRightOuter || inlhjRightOuter || inlmjRightOuter
 	needForced := forceLeftOuter || forceRightOuter
 
+	// These variables are used to enforce the hint. For example, if we have a IndexHashJoin hint
+	// but we cannot generate IndexHashJoin because of the required property, we set `XXXHashJoinMatchProp` = false.
+	// And we will enforce a sort above the join and try to get the IndexHashJoin again.
+	// Because IndexJoin and IndexHashJoin can satisfy the same property, so we use the variable `xxxxHashJoinMatchProp` for them.
+	// IndexMergeJoin has stricter condition, so it uses a separate variable `xxxxMergeJoinMatchProp`.
+	//
+	// In fact, we can simply check whether `prop.IsEmpty` when a hint is inapplicable, then enforce that property and try to apply
+	// the hint again. But this may introduce some useless search. For example, if the join's inner child is not DataSource, it cannot
+	// use IndexJoin whatever the required property is empty or not.
+	leftOuterHashJoinMatchProp, rightOuterHashJoinMatchProp := true, true
+	leftOuterMergeJoinMatchProp, rightOuterMergeJoinMatchProp := true, true
```

Comment:
This might be too complicated. We can simplfy it to `if (needForce && !canForce &&  !prop.IsEmpty) then enforce the property`. But we have a trade off here:
1. If we use the simple method, we may enforce the property for the hint which is inapplicable even if there is not property. For example, if the inner child is not DataSource, the hint cannot work anyway.
2. If we use the complicated method, it looks too ugly and is difficult to understand..

Reply:
I'm trying to use the simple method.